packages/asset-swapper/contracts/src/SwapRevertSampler.sol

```diff
@@ -0,0 +1,406 @@
+        // Address of the token which is being sold
+        address sellToken;
+        // Address of the token which is wanted
+        address buyToken;
+        // Data required for the bridge to execute the swap
+        bytes bridgeData;
+        // Callback to retrieve a swap quote.
+        function (address sellToken, address buyToken, bytes memory bridgeData, uint256 sellAmount)
+            external
+            returns (uint256)
+            getSwapQuoteCallback;
+    }
+
+    struct SwapRevertSamplerBuyQuoteOpts {
+        // Address of the token which is being sold
+        address sellToken;
+        // Address of the token which is wanted
+        address buyToken;
+        // Data required for the bridge to execute the SELL_TOKEN->BUY_TOKEN swap
+        bytes sellTokenData;
+        // Data required for the bridge to execute the BUY_TOKEN->SELL_TOKEN swap
+        bytes buyTokenData;
+        // Callback to retrieve a swap quote.
+        function (address sellToken, address buyToken, bytes memory bridgeData, uint256 sellAmount)
+            external
+            returns (uint256)
+            getSwapQuoteCallback;
+    }
+
+    function _callRevert(
+        bytes4 selector,
+        address sellToken,
+        address buyToken,
+        bytes memory bridgeData,
+        uint256 amountIn
+    )
+        external
+    {
+        // Clear any registered overhead
+        try
+            GasOverhead(GAS_OVERHEAD).clearOverhead()
+        { } catch { }
+        // Measure the gas
+        uint256 gasUsed = gasleft();
+        // Perform the sell
+        (bool success, bytes memory data) = address(this).call(
+            abi.encodeWithSelector(selector, sellToken, buyToken, bridgeData, amountIn)
+        );
+        gasUsed = gasUsed - gasleft();
+        // Remove any registered gas overhead
+        try
+            GasOverhead(GAS_OVERHEAD).overhead()
+            returns (uint256 gasOverhead)
+        {
+            gasUsed = gasUsed - gasOverhead;
+        } catch { }
+
+        if (!success) {
+            data.rrevert();
+        }
+        // Revert with the amount bought
+        _revertSingleSwapSample(abi.decode(data, (uint256)), gasUsed);
+    }
+
+    /// @dev  Mints the sell token, then performs the swap, then reverts with the amount out.
+    /// The SwapRevertSamplerQuoteOpts has been unrolled here as our ABI encoder cannot support
+    /// encoding the function
+    function _mintCallRevert(
+        bytes4 selector,
+        address sellToken,
+        address buyToken,
+        bytes memory bridgeData,
+        uint256[] memory amountsIn
+    )
+        external
+    {
+        IHackedERC20 hackedSellToken = IHackedERC20(payable(sellToken));
+        // We assume the amounts are ascending and that
+        // the underlying call can handle selling a specific amount
+        uint256 amountIn = amountsIn[amountsIn.length - 1];
+
+        // Enable sell token to be tracked and shadowed
+        try
```

Comment:
Yeah I think prefix is the go here, it's still a bit hit and miss when to raise it all the way back to the dev. Since it could be the 8th sample which reverts and the previous 7 are all valid (i.e lack of liquidity in buys), or Balancer which seems to randomly revert on low amounts.